runtime/src/accounts_db.rs

```diff
@@ -681,7 +686,8 @@ impl AccountsDB {
                         let (slot, account_info) = &list[index];
                         if account_info.lamports == 0 {
                             purges.insert(*pubkey, accounts_index.would_purge(pubkey));
-                        } else if accounts_index.uncleaned_roots.contains(slot) {
```

Comment:
Thanks for fixing another AccountsDB bug.. ;)

iirc, there was some reason that I intentionally chained with `else if`. But I can't remember the exact reason. And, sorry for the lack of comment. I'll try to remember and report back here. Anyway, a lot happened since then. So, this is just ok as long as the ci is green and running validator causes no bad snapshot.

Reply:
@ryoqun, no worries! Was it to avoid deleting the slot such that an AppendVec added earlier to `purges` wouldn't be deleted, i.e. so this wouldn't panic? https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L720. I suspect this might be the reason because if you change the `else if` to an `if` on master, and try running this test: https://github.com/solana-labs/solana/pull/12606/files#diff-2099c5256db4eb5975c8834af38f6456R3118, that unwrap() will panic

Reply:
(for the record) this line is added #8436.

Reply:
@carllin Hmm, it seems some of other tests are failing. Could you look at it? Those failing tests will be very cryptic to you. So don't hesitate to ask questions. At first glance, I can't tell these are actually failing by being broken or these test assertions need to just be updated.

I'm still trying to remember something. Still, the memory is blur but it was like this:

- `clean_old_rooted_accounts` should conceptually be complement to the cleaning of zero lamport accounts. So, for every given live account, we do either `clean_old_rooted_accounts` or cleaning of zero lamport accounts. Even if we don't clean old_rooted accounts for zero lamport accounts, the cleaning of zero lamport logic can purge them eventually. So there is no need to do both; it's kind of waste doing both for each account every time we run `clean_accounts()`.
- There was a problem / bug when we're too eager to clean old index entries of zero lamport accounts. So combined the above bullet point concern, I've avoided doing both and chose `if else` approach. (That's why tests failing I guess).
  - ... And I failed to comment as such because that assumption was so prevalent at the time. It was obvious thing to me. (my bad).

Reply:
> @ryoqun, no worries! Was it to avoid deleting the slot such that an AppendVec added earlier to `purges` wouldn't be deleted, i.e. so this wouldn't panic? https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L720. I suspect this might be the reason because if you change the `else if` to an `if` on master, and try running this test: https://github.com/solana-labs/solana/pull/12606/files#diff-2099c5256db4eb5975c8834af38f6456R3118, that unwrap() will panic

As said in the my above comment, I don't think my thought for the original implementation decision started from these concerns. At the time, I didn't even tried to do `if` approach. Anyway, I think your observation is correct for the interrelation of two cleaning and panic of wrap with the `if` approach..

Reply:
> @carllin Hmm, it seems some of other tests are failing. Could you look at it? Those failing tests will be very cryptic to you. So don't hesitate to ask questions. At first glance, I can't tell these are actually failing by being broken or these test assertions need to just be updated.
> 
> I'm still trying to remember something. Still, the memory is blur but it was like this:
> 
> * `clean_old_rooted_accounts` should conceptually be complement to the cleaning of zero lamport accounts. So, for every given live account, we do either `clean_old_rooted_accounts` or cleaning of zero lamport accounts. Even if we don't clean old_rooted accounts for zero lamport accounts, the cleaning of zero lamport logic can purge them eventually. So there is no need to do both; it's kind of waste doing both for each account every time we run `clean_accounts()`.
> * There was a problem / bug when we're too eager to clean old index entries of zero lamport accounts. So combined the above bullet point concern, I've avoided doing both and chose `if else` approach. (That's why tests failing I guess).
>   
>   * ... And I failed to comment as such because that assumption was so prevalent at the time. It was obvious thing to me. (my bad).

Yeah, working on those tests! I think the ones that are failing are a mix of assertions failing and me not handling: https://github.com/solana-labs/solana/pull/12606#discussion_r498093305, which is breaking the issue described here: https://github.com/solana-labs/solana/pull/7157

Reply:
I see! It looks that you passed the CI fully. So happy. :)

So, the once independent cleaning mechanisms finally start to interact each other in one pass after battle-tested for half a year. That sounds cool.